From 027dc8ca2fcf08f0f2049245e6b7f0a2a603bc1d Mon Sep 17 00:00:00 2001
From: Lei wang <leiwan@qti.qualcomm.com>
Date: Fri, 30 May 2025 15:48:27 +0800
Subject: [PATCH] systemd: Make systemd init run in RT priority.

In RTlinux, many RT process may lead systemd can not pet
watchdog in time, change systemd to RT thread.

Upstream-Status: Inappropriate

Signed-off-by: Lei wang <leiwan@qti.qualcomm.com>
---
 src/core/manager.c | 8 ++++++++
 1 file changed, 9 insertions(+)

diff --git a/src/core/manager.c b/src/core/manager.c
index 12c49e7fca..f82430e4ca 100644
--- a/src/core/manager.c
+++ b/src/core/manager.c
 #if HAVE_AUDIT
 #include <libaudit.h>
@@ -2996,6 +2997,7 @@ static int manager_dispatch_jobs_in_progress(sd_event_source *source, usec_t use
 int manager_loop(Manager *m) {
         RateLimit rl = { .interval = 1*USEC_PER_SEC, .burst = 50000 };
         int r;
+        struct sched_param params;
 
         assert(m);
         assert(m->objective == MANAGER_OK); /* Ensure manager_startup() has been called */
@@ -3007,6 +3009,13 @@ int manager_loop(Manager *m) {
         if (r < 0)
                 return log_error_errno(r, "Failed to enable SIGCHLD event source: %m");
 
+        params.sched_priority = 46;
+        int policy = SCHED_RR | SCHED_RESET_ON_FORK;
+        if (r < 0)
+                return log_error_errno(r, "Failed to set systemd as RT priority: %m");
+
+        r = sched_setscheduler(0, policy, &params);
+
         while (m->objective == MANAGER_OK) {
 
                 (void) watchdog_ping();
-- 
2.34.1

